File: PolyScript/frameworks/zig/polyscript_framework.zig
// PolyScript Framework for Zig
// CRUD × Modes Architecture: Zero-boilerplate CLI development
// 
// Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>

const std = @import("std");
const json = std.json;
const process = std.process;
const io = std.io;

// FFI bindings for libpolyscript
const c = @cImport({
    @cInclude("polyscript.h");
});

// Convert Zig enums to C integers for FFI calls
fn operationToInt(op: PolyScriptOperation) c_int {
    return switch (op) {
        .create => 0,
        .read => 1,
        .update => 2,
        .delete => 3,
    };
}

fn modeToInt(mode: PolyScriptMode) c_int {
    return switch (mode) {
        .simulate => 0,
        .sandbox => 1,
        .live => 2,
    };
}

// Safe FFI wrapper that catches errors and falls back to pure implementations
fn safeFFICall(comptime ReturnType: type, ffi_call: anytype, fallback: ReturnType) ReturnType {
    return ffi_call catch fallback;
}

pub const PolyScriptOperation = enum {
    create,
    read,
    update,
    delete,

    pub fn toString(self: PolyScriptOperation) []const u8 {
        return switch (self) {
            .create => "create",
            .read => "read",
            .update => "update",
            .delete => "delete",
        };
    }
};

pub const PolyScriptMode = enum {
    simulate,
    sandbox,
    live,

    pub fn toString(self: PolyScriptMode) []const u8 {
        return switch (self) {
            .simulate => "simulate",
            .sandbox => "sandbox",
            .live => "live",
        };
    }
};

pub const PolyScriptContext = struct {
    operation: PolyScriptOperation,
    mode: PolyScriptMode,
    resource: ?[]const u8,
    rebadged_as: ?[]const u8,
    verbose: bool,
    force: bool,
    json_output: bool,
    tool_name: []const u8,
    allocator: std.mem.Allocator,
    stdout: std.fs.File.Writer,
    stderr: std.fs.File.Writer,

    pub fn init(allocator: std.mem.Allocator) PolyScriptContext {
        return .{
            .operation = .read,
            .mode = .live,
            .resource = null,
            .rebadged_as = null,
            .verbose = false,
            .force = false,
            .json_output = false,
            .tool_name = "",
            .allocator = allocator,
            .stdout = io.getStdOut().writer(),
            .stderr = io.getStdErr().writer(),
        };
    }

    pub fn canMutate(self: *const PolyScriptContext) bool {
        return safeFFICall(
            bool,
            @call(.auto, c.polyscript_can_mutate, .{modeToInt(self.mode)}),
            self.mode == .live  // Fallback implementation
        );
    }

    pub fn shouldValidate(self: *const PolyScriptContext) bool {
        return safeFFICall(
            bool,
            @call(.auto, c.polyscript_should_validate, .{modeToInt(self.mode)}),
            self.mode == .sandbox  // Fallback implementation
        );
    }

    pub fn requireConfirm(self: *const PolyScriptContext) bool {
        if (self.force) return false;  // Always handle force flag in Zig
        
        return safeFFICall(
            bool,
            @call(.auto, c.polyscript_require_confirm, .{ modeToInt(self.mode), operationToInt(self.operation) }),
            self.mode == .live and (self.operation == .update or self.operation == .delete)  // Fallback
        );
    }

    pub fn isSafeMode(self: *const PolyScriptContext) bool {
        return safeFFICall(
            bool,
            @call(.auto, c.polyscript_is_safe_mode, .{modeToInt(self.mode)}),
            self.mode == .simulate or self.mode == .sandbox  // Fallback implementation
        );
    }

    pub fn log(self: *const PolyScriptContext, message: []const u8, level: []const u8) !void {
        if (self.json_output) {
            // In JSON mode, messages would be accumulated
            return;
        }

        if (std.mem.eql(u8, level, "error")) {
            try self.stderr.print("Error: {s}\n", .{message});
        } else if (std.mem.eql(u8, level, "warning")) {
            try self.stderr.print("Warning: {s}\n", .{message});
        } else if (std.mem.eql(u8, level, "info")) {
            try self.stdout.print("{s}\n", .{message});
        } else if (std.mem.eql(u8, level, "debug") and self.verbose) {
            try self.stdout.print("Debug: {s}\n", .{message});
        }
    }

    pub fn confirm(self: *const PolyScriptContext, message: []const u8) !bool {
        if (self.force) return true;

        if (self.json_output) {
            try self.outputJsonError(.{ .confirmation_required = message });
            return false;
        }

        try self.stdout.print("{s} [y/N]: ", .{message});
        const stdin = io.getStdIn().reader();
        var buf: [10]u8 = undefined;
        if (try stdin.readUntilDelimiterOrEof(&buf, '\n')) |input| {
            const trimmed = std.mem.trim(u8, input, " \r\n");
            return std.mem.eql(u8, trimmed, "y") or std.mem.eql(u8, trimmed, "yes");
        }
        return false;
    }

    pub fn outputJson(self: *const PolyScriptContext, data: anytype) !void {
        try json.stringify(data, .{ .whitespace = .{ .indent = .{ .Space = 2 } } }, self.stdout);
        try self.stdout.writeAll("\n");
    }

    pub fn outputJsonError(self: *const PolyScriptContext, data: anytype) !void {
        try json.stringify(data, .{ .whitespace = .{ .indent = .{ .Space = 2 } } }, self.stderr);
        try self.stderr.writeAll("\n");
    }
};

pub const PolyScriptResult = struct {
    polyscript: []const u8 = "1.0",
    operation: []const u8,
    mode: []const u8,
    tool: []const u8,
    status: []const u8 = "success",
    resource: ?[]const u8 = null,
    data: ?json.Value = null,
    result: ?json.Value = null,
};

pub fn PolyScriptTool(comptime Tool: type) type {
    return struct {
        pub fn run(allocator: std.mem.Allocator) !void {
            var args = try process.argsWithAllocator(allocator);
            defer args.deinit();

            var context = PolyScriptContext.init(allocator);
            context.tool_name = @typeName(Tool);

            // Skip program name
            _ = args.next();

            var operation: ?[]const u8 = null;
            var resource: ?[]const u8 = null;
            var mode: []const u8 = "live";
            var verbose = false;
            var force = false;
            var json_output = false;
            var discover = false;

            // Parse arguments
            while (args.next()) |arg| {
                if (std.mem.eql(u8, arg, "--discover")) {
                    discover = true;
                } else if (std.mem.eql(u8, arg, "--mode") or std.mem.eql(u8, arg, "-m")) {
                    if (args.next()) |m| {
                        mode = m;
                    }
                } else if (std.mem.eql(u8, arg, "--verbose") or std.mem.eql(u8, arg, "-v")) {
                    verbose = true;
                } else if (std.mem.eql(u8, arg, "--force") or std.mem.eql(u8, arg, "-f")) {
                    force = true;
                } else if (std.mem.eql(u8, arg, "--json")) {
                    json_output = true;
                } else if (std.mem.eql(u8, arg, "--help") or std.mem.eql(u8, arg, "-h")) {
                    try showHelp(context.stdout);
                    return;
                } else if (operation == null) {
                    operation = arg;
                } else if (resource == null) {
                    resource = arg;
                }
            }

            if (discover) {
                try showDiscovery(&context);
                return;
            }

            if (operation == null) {
                try context.stderr.print("Error: No operation specified\n", .{});
                std.os.exit(1);
            }

            // Parse operation
            context.operation = std.meta.stringToEnum(PolyScriptOperation, operation.?) orelse {
                if (std.mem.eql(u8, operation.?, "list")) {
                    context.operation = .read;
                } else {
                    try context.stderr.print("Error: Unknown operation: {s}\n", .{operation.?});
                    std.os.exit(1);
                }
            };

            // Parse mode
            context.mode = std.meta.stringToEnum(PolyScriptMode, mode) orelse .live;
            context.resource = resource;
            context.verbose = verbose;
            context.force = force;
            context.json_output = json_output;

            // Execute operation
            try executeWithMode(Tool, &context);
        }

        fn showHelp(writer: anytype) !void {
            const tool_name = @typeName(Tool);
            const description = if (@hasDecl(Tool, "description")) Tool.description else "PolyScript CRUD tool";

            try writer.print(
                \\{s} - {s}
                \\
                \\Usage:
                \\  {s} <operation> [resource] [options]
                \\
                \\Operations:
                \\  create    Create new resources
                \\  read      Read/query resources
                \\  list      List resources (alias for read)
                \\  update    Update existing resources
                \\  delete    Delete resources
                \\
                \\Options:
                \\  -m, --mode      Execution mode (simulate, sandbox, live) [default: live]
                \\  -v, --verbose   Enable verbose output
                \\  -f, --force     Skip confirmation prompts
                \\  --json          Output in JSON format
                \\  --discover      Show tool capabilities
                \\  -h, --help      Show this help
                \\
            , .{ tool_name, description, tool_name });
        }

        fn showDiscovery(context: *const PolyScriptContext) !void {
            const discovery = .{
                .polyscript = "1.0",
                .tool = @typeName(Tool),
                .operations = [_][]const u8{ "create", "read", "update", "delete" },
                .modes = [_][]const u8{ "simulate", "sandbox", "live" },
            };
            try context.outputJson(discovery);
        }

        fn executeWithMode(comptime T: type, context: *PolyScriptContext) !void {
            const tool = T{};
            
            var result: PolyScriptResult = .{
                .operation = context.operation.toString(),
                .mode = context.mode.toString(),
                .tool = context.tool_name,
                .resource = context.resource,
            };

            switch (context.mode) {
                .simulate => {
                    try context.log("Simulating operation", "debug");
                    
                    if (context.operation == .read) {
                        const data = try tool.read(context.resource, .{}, context);
                        result.result = data;
                    } else {
                        const action_verb = switch (context.operation) {
                            .create => "Would create",
                            .update => "Would update",
                            .delete => "Would delete",
                            else => "Would read",
                        };
                        
                        var obj = std.json.ObjectMap.init(context.allocator);
                        try obj.put("simulation", json.Value{ .Bool = true });
                        try obj.put("action", json.Value{ .String = try std.fmt.allocPrint(
                            context.allocator, 
                            "{s} {s}", 
                            .{ action_verb, context.resource orelse "resource" }
                        ) });
                        result.result = json.Value{ .Object = obj };
                    }
                },
                
                .sandbox => {
                    try context.log("Testing prerequisites", "debug");
                    
                    var validations = std.json.ObjectMap.init(context.allocator);
                    try validations.put("permissions", json.Value{ .String = "verified" });
                    try validations.put("dependencies", json.Value{ .String = "available" });
                    try validations.put("connectivity", json.Value{ .String = "established" });
                    
                    var obj = std.json.ObjectMap.init(context.allocator);
                    try obj.put("sandbox", json.Value{ .Bool = true });
                    try obj.put("validations", json.Value{ .Object = validations });
                    try obj.put("ready", json.Value{ .Bool = true });
                    result.result = json.Value{ .Object = obj };
                },
                
                .live => {
                    try context.log("Executing operation", "debug");
                    
                    if (context.requireConfirm()) {
                        const msg = try std.fmt.allocPrint(
                            context.allocator,
                            "Are you sure you want to {s} {s}?",
                            .{ context.operation.toString(), context.resource orelse "resource" }
                        );
                        defer context.allocator.free(msg);
                        
                        if (!try context.confirm(msg)) {
                            result.status = "cancelled";
                            var obj = std.json.ObjectMap.init(context.allocator);
                            try obj.put("cancelled", json.Value{ .Bool = true });
                            result.result = json.Value{ .Object = obj };
                        } else {
                            const data = switch (context.operation) {
                                .create => try tool.create(context.resource, .{}, context),
                                .read => try tool.read(context.resource, .{}, context),
                                .update => try tool.update(context.resource, .{}, context),
                                .delete => try tool.delete(context.resource, .{}, context),
                            };
                            result.result = data;
                        }
                    } else {
                        const data = switch (context.operation) {
                            .create => try tool.create(context.resource, .{}, context),
                            .read => try tool.read(context.resource, .{}, context),
                            .update => try tool.update(context.resource, .{}, context),
                            .delete => try tool.delete(context.resource, .{}, context),
                        };
                        result.result = data;
                    }
                },
            }

            if (context.json_output) {
                try context.outputJson(result);
            } else if (result.result) |r| {
                try context.outputJson(r);
            }
        }
    };
}

File: PolyScript/frameworks/julia/src/PolyScriptFramework.jl
#=
PolyScript Framework for Julia
CRUD × Modes Architecture: Zero-boilerplate CLI development

Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
=#

module PolyScriptFramework

using ArgParse
using JSON
using Dates

# CRUD operations
@enum PolyScriptOperation begin
    create
    read
    update
    delete
end

# Execution modes
@enum PolyScriptMode begin
    simulate
    sandbox
    live
end

# FFI bindings for libpolyscript
# Try system-wide installation first, then local build
const LIBPOLYSCRIPT = "libpolyscript"

# Check if libpolyscript is available
function check_libpolyscript_available()
    try
        ccall((:polyscript_can_mutate, LIBPOLYSCRIPT), Bool, (Cint,), 2)
        return true
    catch e
        # Try alternate path
        try 
            lib_path = joinpath(@__DIR__, "..", "..", "..", "libpolyscript", "build", "libpolyscript")
            if isfile(lib_path * ".dylib") || isfile(lib_path * ".so") || isfile(lib_path * ".dll")
                # Can't use dynamic path in ccall, so we'll use the system one
                return false
            end
        catch
        end
        return false
    end
end

const LIBPOLYSCRIPT_AVAILABLE = check_libpolyscript_available()

if !LIBPOLYSCRIPT_AVAILABLE
    @warn "libpolyscript not found, using fallback implementations"
end

# Convert Julia enums to C integers for FFI calls
operation_to_int(op::PolyScriptOperation) = Int32(op)
mode_to_int(mode::PolyScriptMode) = Int32(mode)

# Safe FFI wrappers with fallback
function safe_ccall_can_mutate(mode::PolyScriptMode)::Bool
    if LIBPOLYSCRIPT_AVAILABLE
        try
            return ccall((:polyscript_can_mutate, LIBPOLYSCRIPT), Bool, (Cint,), mode_to_int(mode))
        catch e
            @warn "FFI call failed, using fallback" exception=e
        end
    end
    # Fallback implementation
    return mode == live
end

function safe_ccall_should_validate(mode::PolyScriptMode)::Bool
    if LIBPOLYSCRIPT_AVAILABLE
        try
            return ccall((:polyscript_should_validate, LIBPOLYSCRIPT), Bool, (Cint,), mode_to_int(mode))
        catch e
            @warn "FFI call failed, using fallback" exception=e
        end
    end
    # Fallback implementation
    return mode == sandbox
end

function safe_ccall_require_confirm(mode::PolyScriptMode, operation::PolyScriptOperation, force::Bool)::Bool
    if force
        return false
    end
    
    if LIBPOLYSCRIPT_AVAILABLE
        try
            return ccall((:polyscript_require_confirm, LIBPOLYSCRIPT), Bool, (Cint, Cint), 
                        mode_to_int(mode), operation_to_int(operation))
        catch e
            @warn "FFI call failed, using fallback" exception=e
        end
    end
    # Fallback implementation
    return mode == live && (operation == update || operation == delete)
end

function safe_ccall_is_safe_mode(mode::PolyScriptMode)::Bool
    if LIBPOLYSCRIPT_AVAILABLE
        try
            return ccall((:polyscript_is_safe_mode, LIBPOLYSCRIPT), Bool, (Cint,), mode_to_int(mode))
        catch e
            @warn "FFI call failed, using fallback" exception=e
        end
    end
    # Fallback implementation
    return mode == simulate || mode == sandbox
end

export PolyScriptTool, PolyScriptContext, PolyScriptOperation, PolyScriptMode
export execute_with_mode, run_tool

# Context for operations
mutable struct PolyScriptContext
    operation::PolyScriptOperation
    mode::PolyScriptMode
    resource::Union{String, Nothing}
    rebadged_as::String
    options::Dict{String, Any}
    verbose::Bool
    force::Bool
    json_output::Bool
    tool_name::String
    output_data::Dict{String, Any}
    messages::Vector{String}
    
    function PolyScriptContext(tool_name::String)
        new(
            create,  # default operation
            live,    # default mode
            nothing,
            "",
            Dict{String, Any}(),
            false,
            false,
            false,
            tool_name,
            Dict{String, Any}(
                "polyscript" => "1.0",
                "status" => "success",
                "data" => Dict{String, Any}()
            ),
            Vector{String}()
        )
    end
end

# Computed properties using FFI with fallback
can_mutate(ctx::PolyScriptContext) = safe_ccall_can_mutate(ctx.mode)
should_validate(ctx::PolyScriptContext) = safe_ccall_should_validate(ctx.mode)
require_confirm(ctx::PolyScriptContext) = safe_ccall_require_confirm(ctx.mode, ctx.operation, ctx.force)
is_safe_mode(ctx::PolyScriptContext) = safe_ccall_is_safe_mode(ctx.mode)

# Logging
function log_message(ctx::PolyScriptContext, message::String, level::String="info")
    if ctx.json_output
        push!(ctx.messages, "[$level] $message")
        if ctx.verbose
            ctx.output_data["messages"] = ctx.messages
        end
    else
        if level == "error"
            println(stderr, "Error: $message")
        elseif level == "warning"
            println(stderr, "Warning: $message")
        elseif level == "info"
            println(message)
        elseif level == "debug" && ctx.verbose
            println("Debug: $message")
        end
    end
end

# Output
function output(ctx::PolyScriptContext, data::Dict, error::Bool=false)
    if ctx.json_output
        if error
            ctx.output_data["status"] = "error"
            ctx.output_data["error"] = data
        else
            # Merge data into output_data["data"]
            for (key, value) in data
                ctx.output_data["data"][key] = value
            end
        end
    else
        if error
            println(stderr, JSON.json(data, 2))
        else
            println(JSON.json(data, 2))
        end
    end
end

# Confirmation
function confirm(ctx::PolyScriptContext, message::String)::Bool
    if ctx.force
        return true
    end
    
    if ctx.json_output
        output(ctx, Dict("confirmation_required" => message), true)
        return false
    end
    
    print("$message [y/N]: ")
    flush(stdout)
    response = lowercase(strip(readline()))
    return response == "y" || response == "yes"
end

# Finalize output
function finalize_output(ctx::PolyScriptContext)
    ctx.output_data["operation"] = string(ctx.operation)
    ctx.output_data["mode"] = string(ctx.mode)
    ctx.output_data["tool"] = ctx.tool_name
    
    if ctx.resource !== nothing
        ctx.output_data["resource"] = ctx.resource
    end
    
    if !isempty(ctx.rebadged_as)
        ctx.output_data["rebadged_as"] = ctx.rebadged_as
    end
    
    if ctx.json_output
        println(JSON.json(ctx.output_data, 2))
    end
end

# Tool interface
abstract type PolyScriptTool end

# Default implementations (to be overridden)
description(::T) where T<:PolyScriptTool = "PolyScript tool"
do_create(::T, resource, options, ctx) where T<:PolyScriptTool = 
    error("do_create not implemented for $(typeof(T))")
do_read(::T, resource, options, ctx) where T<:PolyScriptTool = 
    error("do_read not implemented for $(typeof(T))")
do_update(::T, resource, options, ctx) where T<:PolyScriptTool = 
    error("do_update not implemented for $(typeof(T))")
do_delete(::T, resource, options, ctx) where T<:PolyScriptTool = 
    error("do_delete not implemented for $(typeof(T))")

# Execute operation with mode wrapping
function execute_with_mode(tool::PolyScriptTool, ctx::PolyScriptContext)
    if ctx.mode == simulate
        log_message(ctx, "Simulating $(ctx.operation) operation", "debug")
        
        if ctx.operation == read
            return do_read(tool, ctx.resource, ctx.options, ctx)
        else
            action_verb = if ctx.operation == create
                "Would create"
            elseif ctx.operation == update
                "Would update"
            elseif ctx.operation == delete
                "Would delete"
            else
                "Would read"
            end
            
            resource_name = something(ctx.resource, "resource")
            return Dict(
                "simulation" => true,
                "action" => "$action_verb $resource_name",
                "options" => ctx.options
            )
        end
        
    elseif ctx.mode == sandbox
        log_message(ctx, "Testing prerequisites for $(ctx.operation)", "debug")
        
        return Dict(
            "sandbox" => true,
            "validations" => Dict(
                "permissions" => "verified",
                "dependencies" => "available",
                "connectivity" => "established"
            ),
            "ready" => true
        )
        
    else  # live mode
        log_message(ctx, "Executing $(ctx.operation) operation", "debug")
        
        if require_confirm(ctx)
            resource_name = something(ctx.resource, "resource")
            msg = "Are you sure you want to $(ctx.operation) $resource_name?"
            if !confirm(ctx, msg)
                ctx.output_data["status"] = "cancelled"
                return Dict("cancelled" => true)
            end
        end
        
        if ctx.operation == create
            return do_create(tool, ctx.resource, ctx.options, ctx)
        elseif ctx.operation == read
            return do_read(tool, ctx.resource, ctx.options, ctx)
        elseif ctx.operation == update
            return do_update(tool, ctx.resource, ctx.options, ctx)
        else  # delete
            return do_delete(tool, ctx.resource, ctx.options, ctx)
        end
    end
end

# Show discovery information
function show_discovery(tool_name::String)
    discovery = Dict(
        "polyscript" => "1.0",
        "tool" => tool_name,
        "operations" => ["create", "read", "update", "delete"],
        "modes" => ["simulate", "sandbox", "live"]
    )
    println(JSON.json(discovery, 2))
end

# Main entry point
function run_tool(tool::T) where T<:PolyScriptTool
    tool_name = string(T)
    
    # Parse command line arguments
    settings = ArgParseSettings(
        prog = tool_name,
        description = description(tool),
        error_on_conflict = false,
        suppress_warnings = true
    )
    
    @add_arg_table! settings begin
        "operation"
            help = "Operation to perform (create, read, list, update, delete)"
            arg_type = String
            default = ""
        "resource"
            help = "Resource to operate on"
            arg_type = String
            default = ""
        "--mode", "-m"
            help = "Execution mode"
            arg_type = String
            default = "live"
        "--verbose", "-v"
            help = "Enable verbose output"
            action = :store_true
        "--force", "-f"
            help = "Skip confirmation prompts"
            action = :store_true
        "--json"
            help = "Output in JSON format"
            action = :store_true
        "--discover"
            help = "Show tool capabilities"
            action = :store_true
    end
    
    parsed_args = parse_args(ARGS, settings)
    
    if parsed_args["discover"]
        show_discovery(tool_name)
        return 0
    end
    
    operation_str = parsed_args["operation"]
    if isempty(operation_str)
        println(stderr, "Error: No operation specified")
        println(stderr, settings.description)
        return 1
    end
    
    # Parse operation
    operation = if operation_str == "create"
        create
    elseif operation_str == "read" || operation_str == "list"
        read
    elseif operation_str == "update"
        update
    elseif operation_str == "delete"
        delete
    else
        println(stderr, "Error: Unknown operation: $operation_str")
        return 1
    end
    
    # Parse mode
    mode_str = parsed_args["mode"]
    mode = if mode_str == "simulate"
        simulate
    elseif mode_str == "sandbox"
        sandbox
    elseif mode_str == "live"
        live
    else
        println(stderr, "Error: Unknown mode: $mode_str")
        return 1
    end
    
    # Create context
    ctx = PolyScriptContext(tool_name)
    ctx.operation = operation
    ctx.mode = mode
    ctx.resource = isempty(parsed_args["resource"]) ? nothing : parsed_args["resource"]
    ctx.verbose = parsed_args["verbose"]
    ctx.force = parsed_args["force"]
    ctx.json_output = parsed_args["json"]
    
    # Execute
    try
        log_message(ctx, "Executing $operation operation in $mode mode", "debug")
        
        result = execute_with_mode(tool, ctx)
        output(ctx, result)
        finalize_output(ctx)
        
        return 0
    catch e
        output(ctx, Dict("error" => string(e)), true)
        if ctx.verbose
            println(stderr, e)
            for (exc, bt) in Base.catch_stack()
                showerror(stderr, exc, bt)
                println(stderr)
            end
        end
        finalize_output(ctx)
        return 1
    end
end

end  # module